dashboard/app/models/datablock_storage_table.rb

```diff
@@ -0,0 +1,263 @@
+# == Schema Information
+#
+# Table name: datablock_storage_tables
+#
+#  project_id      :integer          not null, primary key
+#  table_name      :string(700)      not null, primary key
+#  columns         :json
+#  is_shared_table :string(700)
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#
+# Indexes
+#
+#  index_datablock_storage_tables_on_project_id  (project_id)
+#
+class DatablockStorageTable < ApplicationRecord
+  # Stores student-owned tables for App Lab's data features, see datablock_storage_controller.rb
+  # Row data for each table is stored in DatablockStorageRecord
+
+  self.primary_keys = :project_id, :table_name
+  has_many :records,
+    -> {order(record_id: :asc)},
+    autosave: true,
+    class_name: 'DatablockStorageRecord',
+    foreign_key: [:project_id, :table_name],
+    dependent: :delete_all
+
+  after_initialize -> {self.columns ||= ['id']}, if: :new_record?
+
+  class StudentFacingError < StandardError
+    attr_reader :type
+
+    def initialize(type=nil)
+      super
+      @type = type
+    end
+  end
+
+  SHARED_TABLE_PROJECT_ID = 0
+
+  # TODO: #57003, implement enforcement of MAX_TABLE_COUNT, we already have
+  # a test for it but we're skipping it until this is implemented.
+  MAX_TABLE_COUNT = 10
+
+  # TODO: #57002, enforce MAX_TABLE_ROW_COUNT, we already have a test for it
+  # but we're skipping it until this is implemented.
+  MAX_TABLE_ROW_COUNT = 20000
+
+  def self.get_table_names(project_id)
+    DatablockStorageTable.where(project_id: project_id).pluck(:table_name)
+  end
+
+  def self.get_shared_table_names
+    get_table_names(SHARED_TABLE_PROJECT_ID)
+  end
+
+  def self.find_shared_table(table_name)
+    DatablockStorageTable.find_by(project_id: SHARED_TABLE_PROJECT_ID, table_name: table_name)
+  end
+
+  def self.add_shared_table(project_id, table_name)
```

Comment:
to check my understanding, this copies a shared table into a project right?

Reply:
It effectively adds a copy-on-write reference to the shared table into the project.  For read operations (see ln 80 `read_records`), if `is_shared_table` is set, we read from the shared table instead (note `is_shared_table` contains the name of the shared table, not a bool - naming was hard for this one).  The first write operation on the table will copy the table into the project and remove the reference (clear `is_shared_table`) (see refs of `if_shared_table_copy_on_write`)